salt/_modules/metalk8s_etcd.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+'''
+Module for handling etcd client specific calls.
+'''
+import logging
+
+from salt.exceptions import CommandExecutionError
+from salt.utils.decorators import depends
+
+PYTHON_ETCD_PRESENT = False
+try:
+    import etcd3
+    PYTHON_ETCD_PRESENT = True
+except ImportError:
+    pass
+
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'metalk8s_etcd'
+
+
+def __virtual__():
+    return __virtualname__
+
+
+def deps_missing(*args, **kwargs):
+    raise CommandExecutionError("python etcd client is not installed")
+
+
+def _get_endpoint_up(ca_cert, cert_key, cert_cert):
+    """Find an etcd endpoint among all servers declared in mine"""
+    endpoints = __salt__['saltutil.runner'](
+        'mine.get',
+        tgt="*",
+        fun='control_plane_ip'
+    ).values()
+
+    for endpoint in endpoints:
+        etcd = etcd3.client(
+            host=endpoint,
+            ca_cert=ca_cert,
+            cert_key=cert_key,
+            cert_cert=cert_cert,
+        )
+
+        try:
+            etcd.status()
+            etcd.close()
+        except etcd3.exceptions.ConnectionFailedError:
+            pass
+        else:
+            return endpoint
+    raise Exception('Fail to find an etcd srv up in the cluster')
+
+
+@depends(PYTHON_ETCD_PRESENT, fallback_function=deps_missing)
+def add_etcd_node(
+        host,
+        endpoint=None,
+        ca_cert='/etc/kubernetes/pki/etcd/ca.crt',
+        cert_key='/etc/kubernetes/pki/etcd/salt-master-etcd-client.key',
+        cert_cert='/etc/kubernetes/pki/etcd/salt-master-etcd-client.crt'):
+    '''Add a new `etcd` node into the `etcd` cluster.
+
+    This module is only runnable from the salt-master on the bootstrap node.
+
+    Arguments:
+        host (str): hostname of the new etcd node
+        endpoint (str): host server in the etcd cluster
+                        IP is expected, not URL
+    '''
+    if not endpoint:
+        # If we have no endpoint get it from mine
+        endpoint = _get_endpoint_up(
```

Comment:
One can provide a set of endpoints to an `etcd` client. We know on which minions `etcd` is supposed to run. We can fetch their `control_plane_ip`s from the mine. Hence we can construct such connection string.

Reply:
The python-etcd3 class constructor support only one endpoint, so we have to pass it one IP address. I updated the _get_endpoint_up to feed the endpoint only with minions with role `etcd`